### PR TITLE
rgw/pubsub: add S3 compliant API to master zone

### PR DESCRIPTION
**Pull request by yuvalif**
## Purpose
This is for allowing notifications directly from the master zone without need for a dedicated pubsub zone, and multisite.

## Refactoring
```rgw_sync_module_pubsub_rest.cc/h``` was refactored into 3 files:
* ```rgw_sync_module_pubsub_rest.cc/h``` - only the Ceph specific REST API, supported on pubsub zone only
* ```rgw_rest_pubsub.cc/h``` - the S3 compliant pubsub REST API
* ```rgw_rest_pubsub_common.cc/h``` - common code between the above 2

As a small cleanup task, the ``rgw_user(const std::string&)`` ctor was made explicit, as this had the potential of causing issues with the xml encoding functions, and overall a good thing :-)

## New API
Master zone support only S3 compliant API, to avoid conflicts with reserved bucket names, used on the pubsube zone with Ceph specific API.
AWS compliant topic creation API was added, based on: https://docs.aws.amazon.com/sns/latest/api/API_CreateTopic.html

### Teuthology
pubsub tests are passing:
http://pulpito.front.sepia.ceph.com/yuvalif-2019-08-21_13:03:41-rgw:multisite-wip-yuval-pubsub-api2-distro-basic-smithi/
> notes: 
> * the multisite failures are regression issues, unrelated to this change
> * notification push tests are skipped, due to infrastructure issues (see: https://github.com/ceph/teuthology/pull/1292)

### Manual Testing
* install rabbitmq server. e.g. ``sudo dnf install rabbitmq-server``
* not necessary to start/enable it, since the binary is executed as part of the tests. note that if there issues with running the rabbitmq-server, the amqp tests will be skipped
* to execute the multisite tests on you machine you would need some python libraries:
```
pip install nose
pip install xmltodict
pip install boto3
pip install boto
pip install pika
```
* assuming all commands are executed from inside the ``build`` directory
* make sure that ceph is not running on the machine where you are executing the tests. e.g. 
```
../src/mstop.sh c1 && ../src/mstop.sh c2 && ../src/stop.sh
```
* copy the following configuration file for the multisite tests, into a file (e.g. ``test_multi.conf``):
```
[DEFAULT]
num_zonegroups = 1
num_zones = 1
gateways_per_zone = 1
log_file = test-multi.log
log_level = 1
file_log_level = 1
num_ps_zones = 1
checkpoint_retries = 60
checkpoint_delay = 5
reconfigure_delay = 5
no_bootstrap = false
```
* due to teuthology issues, the tests that push notifications to an external endpoint (http/amqp) are disabled. to enable them, edit: ``src/test/rgw/rgw_multi/tests_ps.py`` and change ``skip_push_tests`` (at line 31) to ``False``
* to execute the pubsusb (including notifications) tests, run the following from inside the ``build`` dir:
```
MON=1 OSD=1 MGR=0 MSD=0 RGW_MULTI_TEST_CONF=./test_multi.conf nosetests -s ../src/test/rgw/test_multi.py -m "test_ps_*"
```
* when running for the first time, the ``no_bootstrap`` parameter in ``test_multi.conf`` should be set to ``false``. Before any consequent run, you should change to ``true``
* if you want to change the setup in the conf file, you should first stop ceph, change the conf file, and change the ``no_bootstrap`` parameter back to ``false``
* to disable all tests directed to the ``pubsub`` zone, set ``num_ps_zones`` to zero

### CODE REVIEW

**Comment by yuvalif** [comment]
jenkins render docs

File: src/test/rgw/test_rgw_xml.cc [hunk 1]
```diff
@@ -433,3 +433,31 @@ TEST(TestDecoder, Attributes)
...
+    encode_xml("Name", "hello", &f);
+    encode_xml("Status", (i%2 == 0), &f);
+    f.close_section();
+    encode_xml("Value", i, &f);
+    f.close_section();
+  }
+  f.close_section();
+  std::stringstream ss;
+  f.flush(ss);
+  ASSERT_STREQ(ss.str().c_str(), expected_xml_output);
```
Comment (cbodley) [inline src/test/rgw/test_rgw_xml.cc]: nit: on a failed check, gtest will print the first argument as Expected: and the second as Actual:, so the arguments should be swapped

File: doc/radosgw/notifications.rst [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+Bucket notifications provide a mechanism for sending information out of the radosgw when certain events are happening on the bucket.
+Currently, notifications could be sent to HTTP and AMQP0.9.1 endpoints.
+
+Note, that if the events should be stored in Ceph, in addition, or instead of being pushed to an endpoint, 
+the `PubSub Module`_ should be used instead of the bucket notification mechanism.
+
+A user can create different topics. A topic entity is defined by its user and its name. A
+user can only manage its own topics, and can only associate them with buckets it owns.
+
+In order to send notifications for events for a specific bucket, a notification ientity needs to be created. A
```
Comment (cbodley) [inline doc/radosgw/notifications.rst]: typo `ientity`

File: src/test/rgw/test_rgw_xml.cc (hunk 1 above)
Comment (yuvalif) [reply src/test/rgw/test_rgw_xml.cc]: i think its the opposite. first value is actual and second is expected.
see: https://gist.github.com/yuvalif/5e000a5982ed826c937457b9eb252fff

**Comment by yuvalif** [comment]
jenkins render docs

**Comment by cbodley** [comment]
mostly passed qa in http://pulpito.ceph.com/cbodley-2019-09-18_14:51:41-rgw-wip-cbodley-testing-distro-basic-smithi/

(slow ops failures tracked in https://tracker.ceph.com/issues/41834)

**Comment by yuvalif** [comment]
> mostly passed qa in http://pulpito.ceph.com/cbodley-2019-09-18_14:51:41-rgw-wip-cbodley-testing-distro-basic-smithi/
> 
> (slow ops failures tracked in https://tracker.ceph.com/issues/41834)

the pubsub failures: http://qa-proxy.ceph.com/teuthology/yuvalif-2019-09-19_10:44:49-rgw:multisite-wip-yuval-pubsub-api2-distro-basic-smithi/4317021/teuthology.log
are in the push tests (that i forgot to disable, as they dont work in teuthology).
all other pubsub tests are passing.
some other multisite tests are failing in that log, but this is expected.

**Comment by yuvalif** [comment]
rerun of the pubsub tests (only multisite fails):
http://pulpito.front.sepia.ceph.com/yuvalif-2019-09-19_19:57:03-rgw:multisite-wip-cbodley-testing-distro-basic-smithi/

File: src/rgw/rgw_notify.cc [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+    // TODO: add extra filtering criteria: object size, ToD, metadata, ...
+    return false;
+}
+
+int publish(const req_state* s, 
+        const ceph::real_time& mtime, 
+        const std::string& etag, 
+        EventType event_type,
+        rgw::sal::RGWRadosStore* store) {
+    RGWUserPubSub ps_user(store, s->user->user_id);
```
Comment (Jeegn-Chen) [inline src/rgw/rgw_notify.cc]: My understanding on this line is that if User X uploaded an object, the notification config will be retrieved from User X even if User X is not the owner of the Bucket.
Is it an issue here or I misunderstood something?

**Pull request merged**